Implement autofocus support

bendemboski opened the pull request:
When a modal is shown, its root element is given focus so keyboard events can be handled. This change allows developers to use the HTML5 autofocus attribute to indicate that an input element should be given focus, instead of the root element.

This also involved propagating the autofocus attribute through a couple of layers of for/input elements.

simonihmig commented on addon/components/bs-form-element.js:
```diff
@@ -172,6 +172,15 @@ export default FormGroup.extend({
+   * @type string
+   * @public
+   */
+  autofocus: null,
```
This should be a boolean property according to HTML5 specs. Needs updating the example in the docs as well.

simonihmig commented on tests/integration/components/bs-modal-test.js:
```diff
@@ -228,6 +228,89 @@ test('when modal has a form and the submit button is clicked, the form is submit
   this.$('.modal .modal-footer button[type=submit]').click();
 });
 
+test('modal element is focused when no [autofocus] element is present and fade=false', function (assert) {
```
I think the first two tests ("modal element is focused ...") are not needed, even a bit contra productive in the sense that you should write test for public interfaces and not (private) implementation details. Focusing the modal is an implementation detail, the feature is that pressing the escape key closes the modal. This should be testet, and this test is in place already. So just drop these two tests.

However another test that checks if closing the modal with the escape key works when an input element is autofocused would be nice!

simonihmig commented:
Apart from the two little remarks this one looks pretty good. Thanks for your effort so far! :)

bendemboski commented:
Okay, I've pushed a couple of new commits addressing your feedback. If they look good, give me a :+1: and I'll rebase them so there's a clean git history when this PR gets merged.

bendemboski commented:
Okay, last comment addressed, rebased and ready to merge.

simonihmig commented:
Great work, thanks a lot! :)

simonihmig commented on addon/components/bs-modal.js:
```diff
@@ -407,6 +426,20 @@ export default Ember.Component.extend({
+   * @private
+   */
+  takeFocus() {
+    let focusElement = this.$('[autofocus]').first();
```
@bendemboski Did the autofocus feature actually work for you? The tests were fine, and were all passing using PhantomJS. However they failed in Chrome! And indeed the field was never focused in Chrome. 

Which seems right because the above jQuery selector does not select anything, as the autofocus field is not a child of the modal component with the DOM, because the modal-dialog is whormholed to another part of the DOM. This line should be like this:

        let focusElement = this.get('modalElement').find('[autofocus]');

I have a fix ready, however I was just wondering a) did it work for you?, an b) why tests were passing in PhantomJS but failing in Chrome... strange...

bendemboski replied:
Huh, yeah that's weird, not sure why they are passing in PhantomJS. The functionality works for me because I use the renderInPlace flag to disable the ember-wormhole behavior...
